Follow[ers|ing]: Avoid errors when getting avatar url

obenland opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->

See https://wordpress.org/support/topic/critical-error-on-this-website-on-followerspage-page-1-page-2-works/

## Proposed changes:
<!-- Explain what functional changes your PR includes -->
This PR adds comprehensive unit tests for the Followers and Following table classes in response to a critical error reported on the WordPress support forums where avatar data processing was failing with a TypeError.

### Changes made:
* Added unit tests for the Followers table class (`tests/includes/wp-admin/table/class-test-followers.php`)
* Added unit tests for the Following table class (`tests/includes/wp-admin/table/class-test-following.php`)
* Fixed icon processing in the Following table to use `object_to_uri()` consistently like the Followers table
* Tests specifically cover the `column_username` method with ActivityPub icon objects
* Tests use real `prepare_items()` methods and proper ActivityPub actor creation to verify the complete data flow

### Bug context:
The tests were created in response to [this WordPress support topic](https://wordpress.org/support/topic/critical-error-on-this-website-on-followerspage-page-1-page-2-works/) where users experienced a PHP Fatal TypeError with `ltrim()` when processing avatar data that was an array instead of a string. The tests ensure proper handling of ActivityPub actor icon objects throughout the data processing pipeline.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your change as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

### Running the new tests:
1. Ensure you have the WordPress test environment set up with `npm run env-start`
2. Run the specific new tests:
   ```bash
   npm run env-test -- --filter=test_column_username_with_icon_object
   ```
3. Or run all table tests:
   ```bash
   npm run env-test tests/includes/wp-admin/table/
   ```

### Verifying the fix:
1. The tests create real ActivityPub actors with icon objects (containing `type` and `url` properties)
2. They use the actual `prepare_items()` methods to process this data
3. The tests verify that icon objects are properly converted to URL strings using `object_to_uri()`
4. Tests confirm the final HTML output contains the correct escaped avatar URLs

### Manual testing (optional):
1. Set up followers with ActivityPub actors that have complex icon objects
2. Navigate to WP Admin → Users → Followers page
3. Verify that avatar images display correctly without PHP errors
4. Check that the Following page also processes avatars consistently

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs created from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

<!-- Add a changelog message here -->

Add comprehensive unit tests for Followers and Following table classes with proper ActivityPub icon object handling.

</details>

obenland committed:
Add unit tests for Followers table column_username method

- Add comprehensive test class for WP_Admin\Table\Followers
- Test column_username method with ActivityPub actor icon objects
- Simulate realistic data processing from icon object to URL extraction
- Include proper WordPress screen mocking and setup
- Verify HTML output formatting and escaping

obenland committed:
Complete Followers table test with real prepare_items() integration

- Use Followers::add_follower() to create actual follower relationships
- Mock remote actor metadata using pre_get_remote_metadata_by_actor filter
- Test complete data flow: add_follower → prepare_items → column_username
- Verify ActivityPub icon object processing through object_to_uri()
- Test confirms icon object {type: 'Image', url: '...'} → URL string conversion
- All tests pass with real integration of components

obenland committed:
Add Following table test with real prepare_items() integration

- Create test class for Following table column_username method
- Use Actors::upsert() and Following::follow() for real data setup
- Mock remote actor metadata using pre_get_remote_metadata_by_actor filter
- Test complete data flow: upsert → follow → prepare_items → column_username
- Verify ActivityPub icon object processing through object_to_uri()
- Test confirms icon object {type: 'Image', url: '...'} → URL string conversion
- Mirrors Followers table test structure for consistency

obenland committed:
Fix Following table icon processing and improve test formatting

- Fix Following table to use object_to_uri() for consistent icon processing like Followers table
- Clean up Following test formatting and import organization
- Remove unnecessary test assertions for cleaner test structure

matticbot committed:
Add changelog

Jiwoon-Kim commented:
It would be useful to have the option to configure a default image that is displayed when a remote profile has no avatar.

pfefferle commented on includes/wp-admin/table/class-followers.php:
```diff
@@ -223,7 +223,7 @@ public function prepare_items() {
 			$this->items[] = array(
 				'id'         => $follower->ID,
-				'icon'       => $actor->get_icon()['url'] ?? '',
+				'icon'       => object_to_uri( $actor->get_icon() ?? '' ),
```
oh, nice solution! never thought about using it here!

obenland committed:
Merge branch 'trunk' into add-followers-table-tests

obenland commented:
For posterity, the icon object for the actor mentioned in the support forum looked like this:
```json
  "icon": {
    "url": [
      "https://queef.in/storage/profile.webp"
    ],
    "type": "Image",
    "mediaType": "image/webp"
  },
```

obenland committed:
Add test for icon arrays in Following table

Add unit test to verify that actors with icon objects containing
arrays of URLs are handled correctly by the Following table.
Currently this test will fail as the object_to_uri function doesn't
fully process nested arrays in Image objects.

obenland committed:
Fix object_to_uri to handle Image objects with URL arrays

Update object_to_uri function to properly handle Image objects where
the url property is an array of URLs. This fixes the Following table
display when actors have icon objects with multiple URL fallbacks.

The fix recursively processes Image url arrays to extract the first
URL, matching the ActivityStreams vocabulary specification.

pfefferle commented on includes/functions.php:
```diff
@@ -730,7 +730,8 @@ function object_to_uri( $data ) {
 		case 'Image':
-			$data = $data['url'];
+			// See https://www.w3.org/TR/activitystreams-vocabulary/#dfn-image.
+			$data = object_to_uri( $data['url'] );
```
just wanted to suggest that :)

nice 👍

obenland committed:
Merge branch 'trunk' into add-followers-table-tests
